fms/datasets/arrow.py

```diff
@@ -0,0 +1,107 @@
+import sys
+from collections import UserDict
+
+import pyarrow as pa
+import urllib3
+from pyarrow.fs import FileSystem, FileType, LocalFileSystem, S3FileSystem
+from torch.utils.data import Dataset, IterableDataset
+
+from fms.datasets import DatasetStateDictMixin
+
+
+class _ArrowFileData(UserDict):
+    def __init__(self, fs: FileSystem, path: str):
+        self.fs = fs
+        self.path = path
+
+    def __getitem__(self, idx: int):
+        with self.fs.open_input_file(self.path) as file:
+            reader = pa.ipc.open_file(file)
+            print(self.path, file, idx)
+            return reader.get_batch(idx)
+
+    def __iter__(self):
+        with self.fs.open_input_file(self.path) as file:
+            reader = pa.ipc.open_file(file)
+            for i in range(reader.num_record_batches):
+                yield reader.get_batch(i)
+
+    def __len__(self):
+        with self.fs.open_input_file(self.path) as file:
+            reader = pa.ipc.open_file(file)
+            return reader.num_record_batches
+
+
+class ArrowFilesDataset(IterableDataset, DatasetStateDictMixin):
+    """
+    Creates a dataset from a path to a directory of arrow files, either in
+    S3/COS or a local file system.
+
+    uri: s3://endpoint_host/path/to/files or file:///path/to/files
+    step: for distributed training, use rank+1
```

Comment:
Assuming I'm understanding this correctly, iter() uses self.step as the step interval - presumably to allow for each rank to pull one entry from each contiguous set of N entries. If that's the case, then shouldn't step be the world size rather than the individual rank?

Reply:
ah this points to a bug, will fix and clarify.